contributors/design-proposals/support-union-in-api-server.md

```diff
@@ -0,0 +1,200 @@
+Add Tag Indicating To Replace *Union* Fields in Patch
+=============
+
+Add a tag `patchStrategy:"replaceKey"`. For a given type that has this tag, all keys/fields missing 
+from the request will be cleared when patching the object. For a field presents in the request,
+it will be merged with the live config.
+
+This approach could be used to fix the issues with `kubectl apply` (e.g. [#34292](https://github.com/kubernetes/kubernetes/issues/34292))
+
+The proposal of Full Union is in [kubernetes/community#388](https://github.com/kubernetes/community/pull/388).
+
+| Capability | Supported By This Proposal | Supported By Full Union |
+|---|---|---|---|
+| Auto clear missing fields on patch | X | X |
+| Merge union fields on patch | X | X |
+| Validate only 1 field set on type |  | X |
+| Validate discriminator field matches one-of field |  | X |
+| Support non-union patchKey | X | TBD |
+| Support arbitrary combinations of set fields | X |  |
+
+## Proposed Changes
+
+### APIs
+
+**Scope**:
+
+| Union Type | Supported |
+|---|---|---|
+| non-inlined non-discriminated union | Yes |
+| non-inlined discriminated union | Yes |
+| inlined union with [patchMergeKey](https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#strategic-merge-patch) only | Yes |
+| other inlined union | No |
+
+For the *inlined* fields, we have 2 options:
+
+1) Move the tag to the parent struct's.
+
+2) Add some logic to lookup the metadata in go struct of inline union.
+
+We are leaning more to option 1, because the limitation of option 2 is that the metadata associated with
+the inlined APIs will not be reflected in the OpenAPI schema.
+
+#### Examples
+
+There are some examples using option 1 below.
+
+Example of non-inlined non-discriminated union:
+```go
+type ContainerStatus struct {
+	...
+	// Add patchStrategy:"replaceKey"`
+	State ContainerState `json:"state,omitempty" protobuf:"bytes,2,opt,name=state" patchStrategy:"replaceKey"``
+	...
+}
+```
+
+Example of non-inlined discriminated union:
+The tag contains the name of the discriminator.
+```go
+type DeploymentSpec struct {
+	...
+	// Add patchStrategy:"replaceKey/<discriminator-name>"
+	Strategy DeploymentStrategy `json:"strategy,omitempty" protobuf:"bytes,4,opt,name=strategy" patchStrategy:"replaceKey/type"`
+	...
+}
+```
+
+Example of inlined union with `patchMergeKey` only.
+This case is special, because `Volumes` already has a tag `patchStrategy:"merge"`.
+We change the tag to `patchStrategy:"merge|replaceKey"`
+```go
+type PodSpec struct {
+	...
+	// Add another value "replaceKey" to patchStrategy
+	Volumes []Volume `json:"volumes,omitempty" patchStrategy:"merge|replaceKey" patchMergeKey:"name" protobuf:"bytes,1,rep,name=volumes"`
```

Comment:
If we want to use `replaceKeys` in this case, then it will has additional meaning for list of non-primitives, merging list+ union.
I'm leaning to use `merge|replaceKeys` to keep the meaning of each tag clear.